pipe/pipe.py

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python3
+
+from operator import sub
+import os
+from sre_constants import SUCCESS
+import subprocess
+from sys import stdout
+from bitbucket_pipes_toolkit import Pipe, get_logger
+import yaml
+
+logger = get_logger()
+schema = {
+    'AWS_ACCESS_KEY_ID': {'type': 'string', 'required': True},
+    'AWS_SECRET_ACCESS_KEY': {'type': 'string', 'required': True},
+    'CFN_ROLE': {'type': 'string', 'required': False},
+    'STAGE': {'type': 'string', 'required': False},
+    'YARN': {'type': 'boolean', 'required': False},
+    'DEBUG': {'type': 'boolean', 'required': False}
+}
+
+
+class ServerlessDeploy(Pipe):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Deployment Configuration
+        self.access_key_id = self.get_variable('AWS_ACCESS_KEY_ID')
+        self.secret_acces_key = self.get_variable('AWS_SECRET_ACCESS_KEY')
+        self.cfn_role = self.get_variable('CFN_ROLE')
+        self.stage = self.get_variable('STAGE')
+        self.yarn = self.get_variable('YARN')
+
+        # Bitbucket Configuration
+        self.bitbucket_workspace = os.getenv('BITBUCKET_WORKSPACE')
+        self.bitbucket_repo_slug = os.getenv('BITBUCKET_REPO_SLUG')
+        self.bitbucket_pipeline_uuid = os.getenv('BITBUCKET_PIPELINE_UUID')
+        self.bitbucket_step_uuid = os.getenv('BITBUCKET_STEP_UUID')
+        self.bitbucket_commit = os.getenv('BITBUCKET_COMMIT')
+        self.bitbucket_branch = os.getenv('BITBUCKET_BRANCH')
+
+    def inject_aws_creds(self):
+        self.log_debug("Configuring AWS Deployment user.")
+
+        # Create aws cofig directory
+        os.mkdir(os.path.expanduser('~/.aws'))
+
+        f = open(os.path.expanduser('~/.aws/credentials'), "a")
+        f.write("[bitbucket-deployer]")
+        f.write(f'aws_access_key_id={self.access_key_id}')
+        f.write(f'aws_secret_access_key={self.secret_acces_key}')
+        f.close()
+
+    def inject_cfn_role(self):
+        if self.cfn_role is None or not self.cfn_role:
+            self.log_debug("No CFN_ROLE found, skipping.") 
+            return
+
+        self.log_debug("Injecting CFN_ROLE")
+
+        with open(f'{os.getcwd()}/serverless.yml', "r+") as file:
+            try:
+                serverless = yaml.load(file, Loader=yaml.BaseLoader)
+
+                # Ensure iam exists in the provider block
+                if "iam" not in serverless["provider"]:
+                    serverless["provider"]["iam"] = {}
+
+                # if a role already exists DO NOT override it
```

Comment:
Do we not want to keep the previous behaviour? 
```"This will be overwritten with ${CFN_ROLE}. Please remove from serverless.yaml"```

Reply:
Good catch, I had re-wrote this a bit because the old functionality was a little confusing but we do want to maintain it so I've reworked it a little now.

Reply:
Also highlighted a bug where it wouldn't always write the file correctly using the same stream. So I've pulled the writing out to a separate stream now too.